reconcile/status_board.py

```diff
@@ -163,51 +174,70 @@ def get_diff(
             )
+            # new product, so it misses also the applications
+            for app_name in desired_product_apps[product_name]:
+                return_list.append(
+                    StatusBoardHandler(
+                        action="create",
+                        status_board_object=create_app(app_name, product),
+                    )
+                )
 
-        for product, apps in diff_result.change.items():
-            for app in apps.desired - apps.current:
+        # existing product, only add/remove applications
+        for product_name, apps in diff_result.change.items():
+            for app_name in apps.desired - apps.current:
                 return_list.append(
                     StatusBoardHandler(
                         action="create",
-                        status_board_object=Application(
-                            name=app,
-                            fullname=f"{product}/{app}",
-                            product=current_products[product],
+                        status_board_object=create_app(
+                            app_name, current_products[product_name]
                         ),
                     )
                 )
-            for app in apps.current - apps.desired:
+            for app_name in apps.current - apps.desired:
                 return_list.append(
                     StatusBoardHandler(
                         action="delete",
                         status_board_object=Application(
-                            name=app,
-                            fullname=f"{product}/{app}",
-                            product=current_products[product],
+                            name=app_name,
+                            fullname=f"{product_name}/{app_name}",
+                            product=current_products[product_name],
                         ),
                     )
                 )
 
-        for product in diff_result.delete.keys():
-            for application in current_products[product].applications or []:
+        # product is deleted entirely
+        for product_name in diff_result.delete.keys():
+            for application in current_products[product_name].applications or []:
                 return_list.append(
                     StatusBoardHandler(action="delete", status_board_object=application)
                 )
             return_list.append(
                 StatusBoardHandler(
-                    action="delete", status_board_object=current_products[product]
+                    action="delete", status_board_object=current_products[product_name]
                 )
             )
         return return_list
 
+    @staticmethod
+    def apply_diff(
+        dry_run: bool, ocm_api: OCMBaseClient, diff: list[StatusBoardHandler]
+    ) -> None:
+        def _sort_func(x: StatusBoardHandler) -> int:
+            if x.status_board_object.__class__.__name__ == Product.__name__:
+                return 0
+            return 1
```

Comment:
Why sort here? Is that because we need to always process `Product` first, then `Application`?

Instead of using class name reflection, how about add a new order field like `priority` to `StatusBoardhandler`, so the sort can just be
```python
sorted(diff, key=lambda h: h.priority)
```

Reply:
Considered this option, but did not want to pollute the data classes with metadata.

Reply:
@janboll type or class checking in polymorphism is an anti pattern, it `Product` is always higher than `Application`, we could give them a method:

```python
class Product(AbstractStatusBoard):
  def get_priority(self) -> int:
    return 0
```

```python
class Application(AbstractStatusBoard):
  def get_priority(self) -> int:
    return 1
```

```python
sorted(diff, key=lambda h: h. status_board_object.get_priority())
```